### ISSUE TITLE
Mutex.ReleaseMutex throwing on non-Windows operating systems

### ISSUE THREAD

**Issue by jaredpar** [dotnet/coreclr#26659]
Context: https://github.com/dotnet/roslyn/issues/37974 and https://github.com/dotnet/roslyn/issues/38420

Roslyn is getting several reports of crashes on non-Windows operating systems when calling [Mutex.ReleaseMutex](https://docs.microsoft.com/en-us/dotnet/api/system.threading.mutex.releasemutex?view=netframework-4.8). The method is throwing an `ApplicationException` indicating that the non-owning thread attempted to dispose of the value. 

Looking though the code where we acquire and release the `Mutex` is synchronous and can be [seen here](https://github.com/dotnet/roslyn/blob/master/src/Compilers/Shared/BuildServerConnection.cs#L132-L181). The code does reside in an `async` method but the `await` operation occurs after this block. I do plan to factor the sync part out to make it clearer but the code is operating synchronously as written.

Is there reason by which `Mutex.ReleaseMutex` would throw an exception? Or perhaps a known issue that we could be running into. A bit stumped on how to proceed with this issue.

**Comment by kouvel** [comment]
It looks like `TryConnectToServerAsync` is called inside the lock and it does an `await`:
- https://github.com/dotnet/roslyn/blob/af9741b029487cec82fc3232c6821a45e9f16d9c/src/Compilers/Shared/BuildServerConnection.cs#L175
- https://github.com/dotnet/roslyn/blob/af9741b029487cec82fc3232c6821a45e9f16d9c/src/Compilers/Shared/BuildServerConnection.cs#L337

So `ReleaseMutex` may occur on a different thread. Depending on what that task does and due to differences in how IO completions are scheduled on Windows versus Unixes, there may be a possibility that it would be more common on Unixes.

**Comment by jaredpar** [comment]
> It looks like TryConnectToServerAsync is called inside the lock and it does an await

True but the calling code does not `await` the result hence it should not change the thread it's executing on. It's just synchronously executing a function that internally uses `await` hence the thread should not change in the caller. 

> So ReleaseMutex may occur on a different thread. 

Don't see how this can happen. The caller never calls `await` in that block so at what point can the thread switch occur?

**Comment by kouvel** [comment]
I believe the thread switch would occur at the second bullet above, as the caller and callee would complete their work at that point. After that task completes, the callee continues execution (maybe on a different thread) including continuing execution of the caller following call to `TryConnectToServerAsync` and then releases the mutex.

**Comment by kouvel** [comment]
Yea I don't see how that could be happening either. Would it be possible to log the acquire thread ID and the release thread ID in the exception when release throws? A crash dump along with all loaded modules (such that it's debuggable on a different machine) may be useful to get more info.

**Comment by jaredpar** [comment]
> Yea I don't see how that could be happening either. Would it be possible to log the acquire thread ID and the release thread ID in the exception when release throws?

Sure I can make those changes. 

> A crash dump along with all loaded modules (such that it's debuggable on a different machine) may be useful to get more info.

This only reproduces on Mac / Linux hence crash dumps aren't going to be easy to come by.

**Comment by praeclarum** [comment]
Since it's a named mutex, could another thread be referencing the same one and therefore is the creator and/or is calling Dispose?

**Comment by Therzok** [comment]
If it helps any bit, haven't seen anything of the sort of Mono.

**Comment by jaredpar** [comment]
> Since it's a named mutex, could another thread be referencing the same one and therefore is the creator and/or is calling Dispose?

There are definitely multiple threads across multiple processes which manipulate the same `Mutex`. That shouldn't be a problem though. To my understanding the only relationships that need to hold are the following:

1. The thread which calls `WaitOne` is the one which must call `ReleaseMutex`
1. The thread which creates a `Mutex` and passes `initiallyOwned: true` must also call `ReleaseMutex` 

It's possible there is a different invariant that needs to hold here that we're missing but I can't find a reference to one anywhere.

**Comment by kouvel** [comment]
Yes the error is claiming that the thread does not own the lock and attempted to release it, I don't see any issue with that here, the wrapper object too is created/disposed on the same thread and there wouldn't be any concurrent accesses to it.
